### PR TITLE
improve: migrate to thiserror

### PR DESCRIPTION
**Pull request by GopherJ**

### CODE REVIEW

**Comment by GopherJ** [comment]
`Fail` is not compatible with std::Error, so I think it's better to migrate to thiserror.

How do you think, @gsquire ?

**Comment by gsquire** [comment]
This is definitely something I have been wanting to move to, so thanks for the patch. Before I merge it, I wanted to get @Ten0's opinion on #66 error handling. What can be improved here to close that out?

**Comment by GopherJ** [comment]
@gsquire please check the latest commit, it should fix: #66 #65

File: src/client.rs [hunk 1]
```diff
@@ -111,23 +116,15 @@ impl SGClient {
...
-        if !res.status().is_success() {
-            Err(SendgridError::RequestNotSuccessful(
-                res.status(),
-                res.text()?,
-            ))
-        } else {
-            Ok(res.text()?)
-        }
+            .send()
+            .and_then(|x| x.error_for_status())?)
```
Comment (Ten0) [inline src/client.rs:127]: When getting an "invalid status", body of the response would be something like:
```json
{"errors":[{"message":"The content type must be a string at least one character in length.","field":"content.0.type","help":"http://sendgrid.com/docs/API_Reference/Web_API_v3/Mail/errors.html#message.content.type"}]}
```
> @gsquire please check the latest commit, it should fix: #66 #65

On the contrary, e4700e8b8640917cdf297c5c5bc23c5dd48de827 removes the ability for the user to extract the body from the error. I think what was done initially with
```rust
#[error("Request failed with StatusCode: `{0:?}, Body: `{1:?}")]
RequestNotSuccessful(reqwest::StatusCode, String),
```
was perfectly good for debugging, though I'm not sure why the `Display` formatters were replaced by `Debug` formatters, and would probably have favored named fields in the enum to make the meaning of the `String` more obvious:
```rust
pub enum SendgridError {
    ...
    RequestNotSuccessful {
        status_code: reqwest::StatusCode,
        body: String,
    },
}
```

File: src/error.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+use std::io;
+
+use reqwest::{self, header::InvalidHeaderValue};
+use thiserror::Error as ThisError;
+
+/// Represents any of the ways that using this library can fail.
+#[derive(ThisError, Debug)]
+pub enum SendgridError {
+    /// The failure was due to some IO error, for example an interrupted network connection.
+    #[error("IO Error: `{0:?}`")]
```
Comment (Ten0) [inline src/error.rs:10]: With regards to error display, aren't we meant to use `Display` formatters here instead of `Debug` formatters?

**Review by Ten0 — CHANGES_REQUESTED** [review]

File: src/error.rs (hunk 2 above)
Comment (GopherJ) [reply src/error.rs:10]: yeah Display is better

File: src/client.rs (hunk 1 above)
Comment (GopherJ) [reply src/client.rs:127]: so you want body

File: src/error.rs (hunk 2 above)
Comment (GopherJ) [reply src/error.rs]: I created an issue at: https://github.com/seanmonstar/reqwest/issues/1043, I think it's better to make it addressed in reqwest.

File: src/client.rs (hunk 1 above)
Comment (Ten0) [reply src/client.rs:127]: yes

Comment (Ten0) [reply src/client.rs:127]: > I created an issue at: seanmonstar/reqwest#1043, I think it's better to make it to be addressed in reqwest.
Wasn't this supposed to be in this conversation?

**Comment by GopherJ** [comment]
@Ten0 pls review the latest commit

**Review by Ten0 — APPROVED** [review]
Looks great :) 
I think I'd keep the previous writing for extracting the body though, as it's more strict than this one with regards to the expected behavior of Sendgrid, but that's minor in any case :)
Thanks!

File: src/client.rs [hunk 3]
```diff
@@ -118,13 +121,18 @@ impl SGClient {
...
             .client
             .post(API_URL)
             .headers(self.headers()?)
             .body(post_body)
-            .send()
-            .and_then(|x| x.error_for_status())?)
+            .send()?;
+
+        if let Err(_) = resp.error_for_status_ref() {
+            return Err(RequestNotSuccessful::new(resp.status(), resp.text()?).into());
```
[the remark this answers is not in the document]
Comment (GopherJ) [reply src/client.rs:132]: @Ten0 client & server error means (>= 400 && < 600), so I believe it's enough.

File: src/client.rs [hunk 4]
```diff
@@ -154,14 +162,19 @@ impl SGClient {
...
             .post(API_URL)
             .headers(self.headers()?)
             .body(post_body)
             .send()
-            .await
-            .and_then(|x| x.error_for_status())?)
+            .await?;
+
+        if let Err(_) = resp.error_for_status_ref() {
+            return Err(RequestNotSuccessful::new(resp.status(), resp.text().await?).into());
```
[the remark this answers is not in the document]
Comment (GopherJ) [reply src/client.rs:174]: same as above

File: src/v3.rs [hunk 5]
```diff
@@ -139,28 +139,39 @@ impl Sender {
...
             .client
             .post(V3_API_URL)
             .headers(headers)
             .body(body)
-            .send()
-            .and_then(|x| x.error_for_status())?)
+            .send()?;
+
+        if let Err(_) = resp.error_for_status_ref() {
+            return Err(RequestNotSuccessful::new(resp.status(), resp.text()?).into());
```
[the remark this answers is not in the document]
Comment (GopherJ) [reply src/v3.rs:171]: same as above

File: src/v3.rs [hunk 6]
```diff
@@ -139,28 +139,39 @@ impl Sender {
...
             .post(V3_API_URL)
             .headers(headers)
             .body(mail.gen_json())
             .send()
-            .await
-            .and_then(|x| x.error_for_status())?)
+            .await?;
+
+        if let Err(_) = resp.error_for_status_ref() {
+            return Err(RequestNotSuccessful::new(resp.status(), resp.text().await?).into());
```
[the remark this answers is not in the document]
Comment (GopherJ) [reply src/v3.rs:151]: same as above

**Comment by GopherJ** [comment]
@gsquire Could you review again to see if you need some changes?

**Review by gsquire — APPROVED** [review]
Thanks again for these changes! I will publish a new minor release shortly.

**Comment by Elrendio** [comment]
Hello !

This PR actually prevented to read the error body as the field `body` in the struct `RequestNotSuccessful` is not public 😞

From this discussion I believe this is unintentional, therefore I've proposed a PR to fix it: #77

This is actually preventing us from migrating from 0.12 to 0.14 🙂

Have a nice day 🙂

**Pull request merged**